contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+Reference Semantics and also lead to unintended behaviour if we modify the values in 1 object, which in turn affects the other objects which have pointers to the same values as well, an example shown in this [article](http://www.drdobbs.com/cpp/optimization-calling-by-value-or-by-refe/232400151).
+
+## Move Semantics
+As mentioned above, 1 major flaw of Value Semantics is that we will create a copy of the value, which can computationally expensive if this value is a large object. However, we may also not want to use Reference Semantics due to its different problems as discussed above. As such, in order to continue using Value Semantics without creating this unnecessary copy, C++ introduced a new mode, Move Semantics.
+
+### Rvalue and Lvalue references
+To understand how Move Semantics work in C++, we first need to know about the distinction between rvalue and lvalue references. 1 simple line to describe the differencce between these 2 is:
+```
+lvalue = rvalue
+```
+where lvalue (left value) is on the left side of the equation, and rvalue (right value) is on the right.
+lvalue basically refers to values that are addressable, whileas rvalue are temporary objects or values which are used only on the right side of an assignment expression. More details and classification of these 2 values can be found [here](http://www.bogotobogo.com/cplusplus/C11/4_C11_Rvalue_Lvalue.php).
+
+### Using rvalue references
+Rvalue references allows us to distinguish an lvalue from an rvalue. In C++11, we can declare an rvalue reference using the `&&` operator:
+```cpp
+int &&rvalue = 55;
+```
+
+We can also convert an lvalue to a rvalue using the `std::move` function:
+```cpp
+int lvalue = 99;
+int &&rvalue2 = std::move(lvalue); // std::move converts an lvalue to a rvalue
+```
+
+We can also do [function overloading](https://en.wikibooks.org/wiki/Computer_Programming/Function_overloading)
+in order to determine whether the parameters given are lvalues or rvalues:
+
+```cpp
+void print(int& lvalue) { // takes an lvalue
+    std::cout << "lvalue method used";
+}
+
+void print(int&& rvalue) { // take a rvalue
+    std::cout << "rvalue method used";
+}
+
+int main() {
+    int x = 5;
+    print(x); // lvalue method used
+    print(10); // rvalue method used.
+}
+```
+
+Using of these functions overload with rvalue parameters, which take on temporary objects, helps us to write more efficient program using Move Semantics!
+
+### Move Semantics
+The main usage of rvalue references is that it allows us to create *move* constructor and *move* assignments, instead of *copy* constructor and *copy* assignments by default. Since rvalues are typically temporary objects, we can just *move* the value instead of *copying* it, thus helping us to save on memory consumption and improving on the performance on our program as well!
```

Comment:
Rvalue references allows us to create *move* constructor**s** and *move* assignments, instead of *copy* constructor**s** and *copy* assignments by default. Since rvalues are typically temporary objects, we can just *move* the value instead of *copying* it, thus **reducing** memory consumption and improving performance.